lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -740,6 +797,88 @@ ParseResult FIRParser::parseOptionalRUW(RUWAttr &result) {
   return success();
 }
 
+ParseResult FIRParser::importAnnotations(const SMLoc &loc,
+                                         StringRef annotationsStr,
+                                         MLIRContext *context,
+                                         StringRef errorMsgParse) {
+
+  auto annotations = llvm::json::parse(annotationsStr);
+  if (auto err = annotations.takeError()) {
+    handleAllErrors(std::move(err), [&](const llvm::json::ParseError &a) {
+      auto diag = emitError(loc, errorMsgParse);
+      diag.attachNote() << a.message();
+    });
+    return failure();
+  }
+
+  llvm::json::Path::Root root;
+  llvm::StringMap<ArrayAttr> annotationMap;
+  if (annotations) {
+    if (!fromJSON(annotations.get(), annotationMap, root, context)) {
+      auto diag = emitError(loc, "Invalid/unsupported annotation format");
+      std::string jsonErrorMessage =
+          "See inline comments for problem area in JSON:\n";
+      llvm::raw_string_ostream s(jsonErrorMessage);
+      root.printErrorContext(annotations.get(), s);
+      diag.attachNote() << jsonErrorMessage;
+      return failure();
+    }
+  }
+
+  for (auto a : annotationMap.keys()) {
+    if (!state.annotationMap[a]) {
+      state.annotationMap[a] = annotationMap[a];
```

Comment:
This is doing a double lookup, once to check the entry and once to insert it.  The insert method has a form that returns an iterator and a bool "insert succeeded" but in this case I'd recommend:

```
auto &entry = state.annotationMap[a];
if (!entry) {
  entry = annotationMap[a];
  continue;
}
```

... work with `entry` and `annotationMap[a]`

Reply:
Nice! Switching to this.